### PR TITLE
support juniper bgp authentication (READY FOR REVIEW)

### PR DESCRIPTION
**Pull request by kidsbear**
Add basic grammar rule to parse security information for BGP authentication. Those rules are used for [this issue](https://github.com/intentionet/support/issues/11), with a link to [ Configuring BGP Route Authentication](https://www.juniper.net/documentation/en_US/junos/topics/topic-map/bgp-authentication.html#jd0e314). Now considering three authentication commands: [authentication-key](https://www.juniper.net/documentation/en_US/junos/topics/reference/configuration-statement/authentication-key-edit-protocols-bgp.html), [authentication-key-chains](https://www.juniper.net/documentation/en_US/junos/topics/reference/configuration-statement/authentication-key-chains-edit-security.html), and [authentication-algorithm](https://www.juniper.net/documentation/en_US/junos/topics/reference/configuration-statement/authentication-algorithm-edit-protocols-bgp.html). Please help to confirm that all needed information is properly parsed and will be able to access them in JAVA code. ([ConfigurationBuilder](https://github.com/batfish/batfish/blob/master/projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java)). Thanks

### CODE REVIEW

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4 [hunk 1]
```diff
@@ -421,11 +482,39 @@ bps_external_router_id
...
+:
+   (
+      VARIABLE
+      | NON_WHITE_SPACE_STRING
+   )
+   |
+   (
+      DOUBLE_QUOTED_STRING
+   )
+;
```
Comment (ratulm) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4]: why is the rule for string_value not simply: 
( 
    DOUBLE_QUOTED_STRING 
    | NON_WHITE_SPACE_STRING 
    | VARIABLE
)

One group and alphabetical.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4 [hunk 2]
```diff
@@ -21,11 +21,63 @@ b_advertise_peer_as
...
+
+b_auth_key_chain
+:
+   AUTHENTICATION_KEY_CHAIN KEY_CHAIN name = string_value
+   (
+      b_description
+      | b_chain_key
+      | b_tolerance
+   )
```
Comment (ratulm) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4]: make them alphabetical

**Comment by kidsbear** [comment]
Store parsed information in BgpGroup, and pass all information to vendorIndependentConfiguration. This PR does not involve how to use the stored information. Please confirm that each information stored in the desired place.

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AuthenticationKey.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  private static final long serialVersionUID = 1L;
+
+  private BgpAuthenticationAlgorithm _algorithm;
+
+  private AuthenticationKey.KeyOption _option;
+
+  private String _secret;
+  
+  private Date _startTime;
```
Comment (kidsbear) [inline projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/AuthenticationKey.java]: Shouldn't store timeStamp as a Java Date object? What should I use and what is the difference? Thanks

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4 [hunk 4]
```diff
@@ -626,6 +686,11 @@ seik_proposal
    )
 ;
 
+seaks_time
+:
+   DEC DASH DEC DASH DEC PERIOD variable (DASH DEC)?
```
Comment (kidsbear) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4]: This rule is awkward to me. I didn't find any defined rule to parse a String like `2011-6-23.20:19:33-0700`, it will be separated into multiple tokens by `-` if I use defined `string` here.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java [hunk 5]
```diff
@@ -247,6 +257,9 @@ private BgpProcess createBgpProcess(RoutingInstance routingInstance) {
       IpBgpGroup ig = e.getValue();
       BgpNeighbor neighbor = new BgpNeighbor(ip, _c);
       neighbor.setVrf(vrfName);
+      neighbor.setAuthAlgorithm(ig.getAuthAlgorithm());
+      neighbor.setAuthKey(ig.getAuthKey());
+      neighbor.setAuthKeyChainName(ig.getAuthKeyChainName());
```
Comment (kidsbear) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java]: Verify the `keyChainName` actually exists in `_authenticationKeyChains` here? What to do if there is no such `keyChainName` defined?

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4 [hunk 6]
```diff
@@ -38,6 +65,7 @@ b_common
    | b_advertise_inactive
    | b_advertise_peer_as
    | b_as_override
+   | b_auth
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_bgp.g4]: This all needs to be refactored for style. In this rule each alternative should correspond to a full keyword (and spelled like the actual token text), not a category. So you should have alternative like `b_authentication_algorithm`, `b_authentication_key`, `b_authentication_key_chain`. It looks like those exist in the definition of `b_auth`, so they should be moved here and renamed accordingly.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4 [hunk 7]
```diff
@@ -512,6 +513,16 @@ s_security
    )
 ;
 
+se_auth_key_chain
+:
+   AUTHENTICATION_KEY_CHAINS KEY_CHAIN name = string
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4]: Is this supposed to be `AUTHENTICATION_KEY_CHAINS` (plural) or `AUTHENTICATION_KEY_CHAIN` (singular)? Either way, the rule should be renamed to `se_authentication_key_chains` or `se_authentication_key_chain` accordingly.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4 (hunk 4 above)
Comment (arifogel) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/flatjuniper/FlatJuniper_security.g4]: Just make a rule in `FlatJuniper_common.g4` like the `Cisco_common.g4` `variable_permissive` that eats up everything up to the `NEWLINE`.

File: projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java [hunk 8]
```diff
@@ -2246,10 +2256,27 @@ public void exitB_advertise_peer_as(B_advertise_peer_asContext ctx) {
     _currentBgpGroup.setAdvertisePeerAs(true);
   }
 
+  public void exitB_auth_algorithm(B_auth_algorithmContext ctx) {
+    if (ctx.AES_128_CMAC_96() != null) {
+      _currentBgpGroup.setAuthAlgorithm(BgpAuthenticationAlgorithm.AES_128_CMAC_96);
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java]: Please rename `authAlgorithm` to `authenticationAlgorithm`

File: projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java [hunk 9]
```diff
@@ -2246,10 +2256,27 @@ public void exitB_advertise_peer_as(B_advertise_peer_asContext ctx) {
...
+    _currentBgpGroup.setAuthKey(ctx.key.getText());
+  }
+
+  public void exitB_auth_key_chain(B_auth_key_chainContext ctx) {
+    _currentBgpGroup.setAuthKeyChainName(ctx.name.getText());
+  }
+
   @Override
   public void exitB_description(B_descriptionContext ctx) {
-    String description = ctx.description().text.getText();
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java]: I don't have a strong opinion here, but why the change? Were you getting a warning?

File: projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java [hunk 10]
```diff
@@ -3292,6 +3319,40 @@ public void exitS_snmp(S_snmpContext ctx) {
   }
 
   @Override
+  public void exitSe_auth_key_chain(Se_auth_key_chainContext ctx) {
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java]: All of this needs to be broken down into multiple rules. Create a context field `_currentAuthenticationKeyChain` that is set upon entry and exit into `se_auth_key_chain`. Then have `exitXXX` rules for the subrules instead of all these `if` statements.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/BgpGroup.java [hunk 11]
```diff
@@ -21,6 +22,12 @@
 
   private Boolean _advertisePeerAs;
 
+  private String _authKey;
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/BgpGroup.java]: Rename `_authXXX` fields to `_authenticationXXX`

File: projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java [hunk 12]
```diff
@@ -224,6 +231,10 @@ private BgpProcess createBgpProcess(RoutingInstance routingInstance) {
         mg.setLocalAs(routingInstanceAs);
       }
     }
+    // Set default authentication algorithm if missing
+    if (mg.getAuthAlgorithm() == null) {
+      mg.setAuthAlgorithm(BgpAuthenticationAlgorithm.HMAC_SHA_1_96);
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java]: Please refactor this to a constant. That is, make a constant called `DEFAULT_BGP_KEY_CHAIN_AUTHENTICATION_ALGORITHM` set to `BgpAuthenticationAlgorithm.HMAC_SHA_1_96`, and use that constant here.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java [hunk 13]
```diff
@@ -247,6 +258,28 @@ private BgpProcess createBgpProcess(RoutingInstance routingInstance) {
       IpBgpGroup ig = e.getValue();
       BgpNeighbor neighbor = new BgpNeighbor(ip, _c);
       neighbor.setVrf(vrfName);
+      String authKeyChainName = ig.getAuthKeyChainName();
+      if (ig.getAuthKeyChainName() != null) {
+        if (!_c.getAuthenticationKeyChains().containsKey(authKeyChainName)) {
+          _w.redFlag(
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java]: Don't use `_w.redFlag` here. Use `VendorConfiguration.undefined()` instead. You will need a line number to refer to with that call, so you may want to store that as a field like `authenticationKeyChainNameLine`.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java [hunk 14]
```diff
@@ -247,6 +258,28 @@ private BgpProcess createBgpProcess(RoutingInstance routingInstance) {
...
+                  + "for neighbor "
+                  + ig.getRemoteAddress());
+          authKeyChainName = null;
+        } else if (ig.getAuthKey() != null) {
+          _w.redFlag(
+              "Both authentication-key and authentication-key-chain specified for neighbor "
+                  + ig.getRemoteAddress());
+        }
+      }
+      if (ig.getAuthAlgorithm() != null && ig.getAuthKey() != null) {
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java]: Remove this. This will always get flagged because of inheritance of the default authentication algorithm in the master group.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java [hunk 15]
```diff
@@ -92,6 +96,8 @@ private static String communityRegexToJavaRegex(String regex) {
 
   private final Map<String, BaseApplication> _applications;
 
+  private final NavigableMap<String, AuthenticationKeyChain> _authenticationKeyChains;
```
Comment (arifogel) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/JuniperConfiguration.java]: This is supposed to be a Juniper-specific structure. Conversion to vendor-independent should occur in `toVendorConfiguration()`

File: projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java (hunk 9 above)
Comment (kidsbear) [reply projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ConfigurationBuilder.java]: No, but I used inline assignment for other fields like tolerance, so I try to make it consistent.

File: projects/batfish/src/main/java/org/batfish/representation/juniper/BgpGroup.java [hunk 16]
```diff
@@ -72,6 +79,15 @@ public final void cascadeInheritance() {
...
       }
+      if (_authenticationAlgorithm == null) {
+        _authenticationAlgorithm = _parent._authenticationAlgorithm;
+      }
+      if (_authenticationKey == null) {
+        _authenticationKey = _parent._authenticationKey;
+      }
+      if (_authenticationKeyChainName == null) {
+        _authenticationKeyChainName = _parent._authenticationKeyChainName;
+      }
```
Comment (kidsbear) [inline projects/batfish/src/main/java/org/batfish/representation/juniper/BgpGroup.java]: One thing I am not sure: what if a BgpGroup specified authentication keychain and algorithm,(by authentication-key-chain and authentication-algorithm commands), and its parent BgpGroup has authentication-key specified, should the child get the `authentication key` from its parent?
